### PR TITLE
Add Support for FEMNIST

### PR DESCRIPTION
**Pull request by SamuelGong**
Add support for the FEMNIST dataset by referring to an open-sourced project [LEAF](https://github.com/TalwalkarLab/leaf).

## Description
Main changes:

1. Added a new datasource `femnist` at `~/plato/datasources/femnist.py` and modified the `~/plato/datasources/registry.py` accordingly.
2. Added a new sampler `empty` at `~/plato/samplers/empty.py` and modified the `~/plato/samplers/registry.py` accordingly.
3. Made minor changes at `~/plato/trainers/basic.py` and `~/plato/models/lenet5.py` for further adaptation.

*Remark: while the implementation is mainly borrowed from LEAF, it does not need to plug in the LEAF project and can work independently.*

Motivation

Apart from the label distribution skew (that can be implemented with LDA), non-IID scenarios also consist of other circumstances including (1) feature distribution skew, (2) same label with different features,  as well as (3) same feature with different labels (see this [survey](https://arxiv.org/abs/1912.04977) for more details). Thus, it would be useful if Plato can supports FL research with more realistic datasets. The FEMNIST dataset is one celebrated example, and it is inherently partitioned by the clients' identification. We thus considered adding support for it, and hopefully, our design can be compatible with those realistic datasets that are also partitioned by clients.

*p.s. One may want to refer to an external [tutorial](https://github.com/SamuelGong/plato/blob/zhifeng-async/zhifeng_learning/exploration/2_customize_datasource/README.md) in a forked version of Plato for more context.*

## How has this been tested?

At the root directory,
```bash
conda activate federated
python run --config=./examples/async/data_hetero/data_hetero_femnist_lenet5.yml > out.txt 2>&1 &
```

*p.s. Please expect hours for the first test in your environment due to the data preprocessing overhead.*

In our test, we observed the generated `out.txt` and confirmed that the training can go on smoothly.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue) Fixes #
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by baochunli** [comment]
I am not sure why you need to include the entire LEAF git repo as a submodule. Your code doesn't seem to depend on this git repo at all (which is excellent). Since Plato is designed as a lightweight framework that can be installed using pip, it would be best to keep it lean and simple, and avoid git submodules all together. If third-party frameworks need to be included, it would be best to add them into requirements.txt for pip install.

**Comment by SamuelGong** [comment]
I see and let me think of implementing the logic for downloading and preprocessing solely with Plato.

**Comment by baochunli** [comment]
> I see and let me think of implementing the logic for downloading and preprocessing solely with Plato.

Sure, thank you. One minor addition: commit messages can have a bit more clarity -- rather than say 'data hetero first commit' it may be better to say 'Supported heterogeneity of data distribution." (with a full stop at the end and with a verb to start the sentence, in past tense).

**Comment by SamuelGong** [comment]
Thanks for the advice and I will take care of my commit messages next time. BTW, the code has been amended now, and please let me know if there is any issue.

**Comment by SamuelGong** [comment]
Just cleaned up the pull request.

**Comment by SamuelGong** [comment]
The PR is ready for review now.

**Comment by baochunli** [comment]
```
from zhifeng_learning.patch.utils import (
    CustomDictDataset, ReshapeListTransform
)
```

This code in `datasources/femnist.py` does not work as the import does not exist. Perhaps you can move the code into `plato/utils`?

Please try to do a fresh clone of the git repo with this PR and test it out first.

**Comment by baochunli** [comment]
I have committed some minor changes to the codebase. However, I am at a loss to understand why you wouldn't use PyTorch (or TensorFlow)'s built-in support for the EMNIST dataset -- which is available as a part of TorchVision:

https://pytorch.org/vision/stable/datasets.html#emnist

Rather than directly using this from TorchVision, you chose to manually download the data, and process + load them manually completely in memory, which is very slow. If I were to do it, I would use EMNIST directly for TorchVision.

**Comment by SamuelGong** [comment]
Hi authors, the new feature has been tested on a fresh clone, with the details noted down [here](https://github.com/SamuelGong/plato/tree/zhifeng-async/zhifeng_learning/exploration/2_customize_datasource#pull-request). The PR is now ready to be reviewed, again.

According to PyTorch's [documentation](https://pytorch.org/vision/stable/datasets.html#emnist), its `emnist` does not have `bywrite` supported in the `split` option of the constructor of the class `torchvision.datasets.EMNIST`. In other words, what we want is the support for datasets which is partitioned by data providers' identification, while we are afraid that PyTorch's `emnist` does not provide this option. More details on the differences between `bywrite` and other options of `split` can be referred to the related [paper](https://ieeexplore.ieee.org/stamp/stamp.jsp?arnumber=7966217).

Please let us know for any further concerns.

**Comment by baochunli** [comment]
All my previous changes to `datasources/femnist.py` were lost -- when you work on the PR, please start from the latest PR commits so that you don't end up overwriting changes from the others.

**Comment by SamuelGong** [comment]
Dear author, thanks for the attention! Actually, we do start from the latest PR commit. We intentionally eliminated most of the previous preprocessing logic. We have not released our changes as we are conducting some necessary tests. We will vocalize the rationale for such a decision after we have settled.

BTW, the last version of `femnist.py` (and thus your efforts put) is being well preserved in a copy [here](https://github.com/SamuelGong/plato/blob/zhifeng-async/plato/datasources/femnist_old.py). In other words, although we think that it is not appropriate for Plato to insist on using the previous idea, we still keep it as a reference point in our tutorial. Hope this releases your concerns :)

**Comment by baochunli** [comment]
This is not a true statement. My changes in the opening comments, for example, have been eliminated. All my changes to `logging.info` have been lost too.

**Comment by SamuelGong** [comment]
> This is not a true statement. My changes in the opening comments, for example, have been eliminated. All my changes to `logging.info` have been lost too.

Oh, my bad. I saw the problem. I have amended the [copy](https://github.com/SamuelGong/plato/blob/zhifeng-async/plato/datasources/femnist_old.py) now.

**Comment by SamuelGong** [comment]
Hi authors, the PR has been amended in terms of both runtime and memory efficiency, so it now is ready for review. More context can be found via an [external tutorial](https://github.com/SamuelGong/plato/tree/zhifeng-async/zhifeng_learning/exploration/3_customize_datasource_cont). Have fun!

**Comment by baochunli** [comment]
Looks to be in great shape. Merging into the main branch.
